pkg/channel/jetstream/dispatcher/consumer.go

```diff
@@ -19,122 +19,226 @@ package dispatcher
 )
 
+func init() {
+	batchSize, err := strconv.Atoi(os.Getenv(batchGoRoutinesEnv))
+	if err == nil {
+		NumBatchGoRoutines = batchSize
+	}
+
+	fetchSize, err := strconv.Atoi(os.Getenv(fetchBatchSizeEnv))
+	if err == nil {
+		FetchBatchSize = fetchSize
+	}
+}
+
 type Consumer struct {
-	sub              Subscription
 	dispatcher       *kncloudevents.Dispatcher
 	reporter         eventingchannels.StatsReporter
 	channelNamespace string
 
-	jsSub *nats.Subscription
+	natsConsumer     jetstream.Consumer
+	natsConsumerInfo *jetstream.ConsumerInfo
 
 	logger *zap.SugaredLogger
-	ctx    context.Context
 
-	mu     sync.Mutex
-	closed bool
+	sub   Subscription
+	subMu sync.RWMutex
+
+	lifecycleMu sync.Mutex
+	started     bool
+	closing     chan struct{}
+	closed      chan struct{}
 }
 
-func (c *Consumer) Close() error {
-	c.mu.Lock()
-	defer c.mu.Unlock()
-	if c.closed {
-		return ErrConsumerClosed
+func NewConsumer(
+	ctx context.Context,
+	consumer jetstream.Consumer,
+	subscription Subscription,
+	dispatcher *kncloudevents.Dispatcher,
+	reporter eventingchannels.StatsReporter,
+	channelNamespace string,
+) (*Consumer, error) {
+	consumerInfo, err := consumer.Info(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get consumer info: %w", err)
 	}
 
-	c.closed = true
+	logger := logging.FromContext(ctx)
+
+	return &Consumer{
+		dispatcher:       dispatcher,
+		reporter:         reporter,
+		channelNamespace: channelNamespace,
+
+		natsConsumer:     consumer,
+		natsConsumerInfo: consumerInfo,
+
+		logger: logger,
+
+		sub: subscription,
 
-	// TODO: should we wait for messages to finish and c.jsSub.IsValid() to return false?
-	return c.jsSub.Drain()
+		closing: make(chan struct{}),
+		closed:  make(chan struct{}),
+	}, nil
 }
 
-func (c *Consumer) MsgHandler(msg *nats.Msg) {
-	logger := c.logger.With(zap.String("msg_id", msg.Header.Get(nats.MsgIdHdr)))
-	ctx := logging.WithLogger(c.ctx, logger)
-	tickerCtx, tickerCancel := context.WithCancel(c.ctx)
+// Start begins the consumer and handles messages until Close is called. This method is blocking and
+// will return an error if the consumer fails prematurely. A nil error will be returned upon being
+// stopped by the Close method.
+func (c *Consumer) Start() error {
+	if err := c.checkStart(); err != nil {
+		return err
+	}
 
-	tickerDone := make(chan struct{})
+	defer close(c.closed)
 
-	go func() {
-		defer close(tickerDone)
+	ctx := logging.WithLogger(context.Background(), c.logger)
 
-		// TODO(dan-j): this should be a fraction of the Consumer's AckWait
-		ticker := time.NewTicker(10 * time.Second)
-		defer ticker.Stop()
+	var wg sync.WaitGroup
 
-		for {
-			select {
-			case <-tickerCtx.Done():
-				return
-			case <-ticker.C:
-				if err := msg.InProgress(nats.Context(tickerCtx)); err != nil && !errors.Is(err, context.Canceled) {
-					logging.FromContext(ctx).Errorw("failed to mark message as in progress", zap.Error(err))
-				}
+	// wait for handlers to finish
+	defer wg.Wait()
```

Comment:
Messages are processed by `consumeMessages` in new goroutines, then we immediately fetch a new batch. We don't wait for existing messages to be handled before fetching a new batch (this was my first approach which my comment on the `NumBatchGoRoutines` constant is referencing, I can add this back in if you wanted to see what I did there).

As such, `wg` is tracking how many messages are being handled by `consumeMessages` in total, and we wait for all these to exit when the consumer is closed.

Reply:
If it does not wait for previous batch to complete before taking the next batch, it will work the same way as push-based consumers, it will consume all cpu fast if there are a lot of messages in a stream.
In case there are other consumers it will affect those as well. We notice high cpu consumption of dispatcher when do load tests. To control back pressure with push based consumers we set lower limits of maxAckPending parameter and control maximum number of in-flight messages.
We need to think about something like this to control back pressure per consumer.